lib/ui/semantics/attributed_string.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/semantics/attributed_string.h"
+
+#include "flutter/fml/logging.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/tonic/dart_args.h"
+#include "third_party/tonic/dart_binding_macros.h"
+
+#include <memory>
+
+namespace flutter {
+
+static void AttributedString_constructor(Dart_NativeArguments args) {
+  UIDartState::ThrowIfUIOperationsProhibited();
+  DartCallConstructor(&AttributedString::Create, args);
+}
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, AttributedString);
+
+#define FOR_EACH_BINDING(V)                 \
+  V(AttributedString, initAttributedString) \
+  V(AttributedString, initWithConcat)       \
+  V(AttributedString, GetString)            \
+  V(AttributedString, toString)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+void AttributedString::RegisterNatives(tonic::DartLibraryNatives* natives) {
+  natives->Register(
+      {{"AttributedString_constructor", AttributedString_constructor, 1, true},
+       FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
+}
+
+AttributedString::AttributedString() {}
+
+AttributedString::~AttributedString() {}
+
+fml::RefPtr<AttributedString> AttributedString::Create() {
+  return fml::MakeRefCounted<AttributedString>();
+}
+
+void AttributedString::initAttributedString(
+    std::string string,
+    std::vector<NativeStringAttribute*> attributes) {
+  string_ = string;
+  for (auto& native_attribute : attributes) {
+    attributes_.push_back(native_attribute->attribute_);
+  }
+}
+
+void AttributedString::initWithConcat(AttributedString* first,
+                                      AttributedString* second) {
+  if (first->string_.empty()) {
+    string_ = second->string_;
+    attributes_ = second->attributes_;
+    return;
+  }
+  if (second->string_.empty()) {
+    string_ = first->string_;
+    attributes_ = first->attributes_;
+    return;
+  }
+
+  // None of the strings is empty.
+  string_ = first->string_ + second->string_;
+  attributes_ = first->attributes_;
+  size_t offset = first->string_.size();
+  for (auto attribute : second->attributes_) {
+    auto attribute_copy = attribute->copy();
+    attribute_copy->start += offset;
+    attribute_copy->end += offset;
+    attributes_.push_back(attribute_copy);
+  }
+  return;
+}
+
+std::string AttributedString::toString() {
```

Comment:
I see no good reason to do this on the native side. If we really want these things to not be opaque, they should just be present on the dart object. Making them present in `toString` will lead to people (and tests) depending on this, which becomes brittle any time we change the string representation of this. Instead tests/developers should be able to assert on properties in the class.